### ISSUE TITLE
Update DotNetCore Aliases to match tooling 1.0

### ISSUE THREAD

**Issue by JonCubed** [cake-build/cake#1533]
With the release of VS 2017,  .NET Core Tooling has been bumped to 1.0. This has introduced a new build system (move from project.json to csproj and msbuild), changes to existing commands and added additional commands. It would be nice for the DotNetCore Aliases to reflect the changes, making it easier for ppl to migrate over to the new tooling.

I asked in gitter and @gep13 said you would be willing to accept a PR on this and I am happy to put some time in to do it. However I have some questions before starting

1. Given that VS 2015 wont use v1.0 of the tooling should we continue to support Project.Json version of the tooling for the time being?
2. If we support both we will need a new set of alias to distinguish, any thoughts on how you prefer for this to be? Does the new tooling become `DotNetCore*` and old become `DotNetCoreJson*` creating a breaking change or keep current as is and new Aliases called something like `DotNetCoreMSBuild*`?
3. Would it be easier for review process if I split this into multiple PRs?

**Tasks:**
- [x] Common Changes - ~#1549~
- [x] Add DotNetCoreClean - ~#1555~
- [x] Add DotNetCoreNugetPush - ~#1581~
- [x] Add DotNetCoreNugetDelete - ~#1565~ 
- [X] Add DotNetCoreMSBuild - ~#1599~
- [x] Add DotNetCoreVSTest - ~#1675~
- [x] Review DotNetCoreRestore - ~#1534~
- [x] Review DotNetCoreBuild - ~#1550~
- [x] Review DotNetCorePublish - ~#1552~
- [x] Review DotNetCoreTest - ~#1554~
- [x] Review DotNetCorePack - ~#1553~
- [x] Review DotNetCoreRun - no change required
- [x] Review DotNetCoreExecute -~#1551~
- [x] Review if any changes required for 1.0.4 of tooling - no change required

**Comment by gep13** [comment]
1. If the "new" way of doing it, is going to be "the" way of doing it, then I would be happy updating the existing aliases to only use that method.  Community Members who require the earlier aliases have the option to pin to an earlier version of Cake.
2. I think this would get messy pretty quickly, and would prefer to not go down this route.
3. Possibly yes, depending on how much changes

@cake-build/team thoughts?

**Comment by patriksvensson** [comment]
@gep13 @JonCubed

1. Agree with @gep13 
2. Agree with @gep13 
3. I think it makes sense to break each tool down to a separate PR since it will be very clear in the commit log and also the release notes.

**Comment by JonCubed** [comment]
@gep13 @patriksvensson 

1. Good to know, that was what I was hoping for.
3. Multiple it is!

**Comment by JonCubed** [comment]
@gep13 @patriksvensson are you able to provide feedback if any on PRs?

I'll move on to Nuget next.

**Comment by patriksvensson** [comment]
@JonCubed I will try to find some time during the day or tomorrow and have a look. 👍

**Comment by devlead** [comment]
@JonCubed I'm starting to review your .NET Core PRs now.
I've changes the target branch for these to `cake-build:feature/dotnetV10tooling`, It's based on todays develop and I'll keep it in sync with develop.

I've assigned a separate milestone for this issue and underlying PRs (just so I can easily find them for now :wink: ) , with the aim to do it's own release with these changes when done.
Having it's own branch lets us merge & close PRs, test and rebase against develop and when done merge into develop and release.

Am I right to assume #1549 (Common changes for DotNetCore Alias) is what we need to review and get merged first?

**Comment by JonCubed** [comment]
Correct #1549 should be merged first

**Comment by devlead** [comment]
@JonCubed I've rebased your PRs now and will start with reviewing the dotnet restore one ( #1534 ).

**Comment by JonCubed** [comment]
@devlead awesome! thanks

**Comment by JonCubed** [comment]
@devlead thanks for reviewing and merging all the PRs. I'll try to get the remaining over the next week

**Comment by JonCubed** [comment]
In Gitter @patriksvensson suggested that xunit .net core tool should be supported by core, should I add it to the list if no one else is working on it?

cc @gep13 @devlead

**Comment by devlead** [comment]
@JonCubed xunit is out of scope for this issue/milestone, will look how we want to add that later.

**Comment by JonCubed** [comment]
@devlead sure no worries

**Comment by gep13** [comment]
@JonCubed @devlead on this subject...

Add DotNetCoreMSBuild
Add DotNetCoreVSTest

Are those extensions to the core dotnet cli tooling, in the same way that xunit is?  If so, I think they are a candidate for dealing with elsewhere as well.

Thoughts?

**Comment by devlead** [comment]
@gep13 yes those are part of the official SDK
```
.NET Command Line Tools (1.0.1)
Usage: dotnet [host-options] [command] [arguments] [common-options]

Arguments:
  [command]             The command to execute
  [arguments]           Arguments to pass to the command
  [host-options]        Options specific to dotnet (host)
  [common-options]      Options common to all commands

Common options:
  -v|--verbose          Enable verbose output
  -h|--help             Show help

Host options (passed before the command):
  -d|--diagnostics      Enable diagnostic output
  --version             Display .NET CLI Version Number
  --info                Display .NET CLI Info

Commands:
  new           Initialize .NET projects.
  restore       Restore dependencies specified in the .NET project.
  build         Builds a .NET project.
  publish       Publishes a .NET project for deployment (including the runtime).
  run           Compiles and immediately executes a .NET project.
  test          Runs unit tests using the test runner specified in the project.
  pack          Creates a NuGet package.
  migrate       Migrates a project.json based project to a msbuild based project.
  clean         Clean build output(s).
  sln           Modify solution (SLN) files.

Project modification commands:
  add           Add items to the project
  remove        Remove items from the project
  list          List items in the project

Advanced Commands:
  nuget         Provides additional NuGet commands.
  msbuild       Runs Microsoft Build Engine (MSBuild).
  vstest        Runs Microsoft Test Execution Command Line Tool.
```

**Comment by gep13** [comment]
@devlead perfect! 👍

**Comment by JonCubed** [comment]
@gep13 most of the original commands are now just wrappers for the advance
commands in the new tooling.

**Comment by devlead** [comment]
@JonCubed FYI I've adjusted the integration tests for the new tooling and rebased branch against latest develop. So you might need to refresh and rebase any PRs.

**Comment by JeremyCade** [comment]
@devlead RE `dotnet xunit` it may pay to have a generic `DotNetCoreTool` alias?

More than happy to put that together.

**Comment by devlead** [comment]
> @devlead RE dotnet xunit it may pay to have a generic DotNetCoreTool alias?
>
> More than happy to put that together.

@JeremyCade Sounds like a reasonable feature to add, would you mind raising a separate issue for this, so we can track/discuss it separately and feel free to do a WIP PR if you want to.

**Comment by JeremyCade** [comment]
> @JeremyCade Sounds like a reasonable feature to add, would you mind raising a separate issue for this, so we can track/discuss it separately and feel free to do a WIP PR if you want to.

Done. See #1577

**Comment by JonCubed** [comment]
@devlead I've added a [PR](#1599) for the initial work on msbuild command. I'm not really familiar with all the different settings of msbuild so I had to rely on documentation descriptions which wasn't always clear. Let me know if you have any feedback on it. 

Anyone else watching this issue is welcome to also provide feedback.

**Comment by RichiCoder1** [comment]
@JonCubed `dotnet msbuild` should functionally be a slight superset of the existing msbuild alias. All it's doing is invoking MSBuild :)

**Comment by patriksvensson** [comment]
Agree with @RichieCoder2. Would be awesome if we could expose the MSBuildSettings on dotnet build.

**Comment by JonCubed** [comment]
@RichiCoder1 @patriksvensson `dotnet msbuild` is running an embedded version of the cross-platform MSBuild 15, I think VS2017 ships with the Full Framework version of MSBuild 15.  There are some similarities with previous versions of MSBuild but there are also some difference such as switches that are no longer supported and other switches that take different values now. 

If the `DotNetCoreMSBuildSettings` was a superset of `MSBuildSettings` we would end up with settings that would do absolutely nothing and be misleading. Also not sure how we would nicely handle properties that would take a different value type, overrides? but this would mean we couldn't just reuse `MSBuildSettings`. I think this way is probably not desirable, but correct me if I'm wrong.

Alternatively we could do some abstraction/base class to share as much commonness as possible not sure how coupled we want these settings to be as they could diverge somewhat going forward.

Another issue is the `DotNetCoreTools` need settings that inherit from `DotNetCoreSettings`, which is why I extracted all the argument building into `MSBuildArgumentBuilderExtensions` so it could be reused else were including `MSBuildRunner` in the future.

The `MSBuild` Aliases is for standalone MSBuild and so it works differently to .NET core version, such as the need for alternative tool paths. We would probably still need to have different `Tool` classes regardless of which way we go. 

Open to ideas on how we might go about merging them together.

**Comment by JonCubed** [comment]
I've given this a bit more thought. I took a look at the command line doc for older versions of msbuild, it hasn't really changed, the `MSBuildSettings` made me think there was a greater difference then there really is. So here is what I propose...

* Create an interface `ICommonMSBuildSettings` that both `MSBuildSettings` and `DotNetCoreMSBuildSettings` implement. 
    * This will solve the issue that `DotNetCoreMSBuildSettings` needs to inherit from `DotNetCoreSettings`.
    * MSBuild alias will be up-to-date with version 15

* All of the common argument logic in `MSBuildRunner` will be replaced with `MSBuildArgumentBuilderExtensions` and `AppendMSBuildSettings` will take `ICommonMSBuildSettings` instead of `DotNetCoreMSBuildSettings`

* Duplicate classes will be merged i.e. `MSBuildFileLoggerSettings` and `MSBuildFileLogger`

* We maintain the 2 tools and aliases `DotNetCoreMSBuild` and `MSBuild`

some decisions will need to be made were there are mismatches

* ToolVersion for `MSBuildRunner` is used to determine MSBuild path, in `DotNetCoreMSBuild` it is used for `/toolversion` switch. They currently use different enums. 
    * We could leave this out of `ICommonMSBuildSettings` and leave each tool using how it does currently?

* Verbosity is set via different enums, `Verbosity` enum has `Verbose` while `DotNetCoreVerbosity` has `Detailed`
    * Leave as is?
    * Change DotNetCoreTools to use `Verbosity` and rename `Verbose` to `Detailed`? 
    * Change DotNetCoreTools to use `Verbosity` and mark `Verbose` obsolete, add `Detailed`
    * Change DotNetCoreTools to use `Verbosity` , DotNetCoreTool handles mismatch

* For duplicate classes, what should we do for classes and where property names are different?
   * It is a breaking change so lets go with the best name?
   * Keep breaking changes to a minimum, use existing name?

**Comment by patriksvensson** [comment]
@JonCubed Excellent research. A lot of base classes shared between two different tools are bound to come back biting us. Perhaps it's better to keep the tools separated like you originally did and merge them together once the whole dotnet thing has stabilized a little.

**Comment by devlead** [comment]
@JonCubed agree with @patriksvensson, I've rebased the the feature branch and #1533 with latest develop. So just finish the oustanding tests/extensions mentioned in the PR and I'll do an review.

**Comment by JonCubed** [comment]
just a quick update, life and work has been a little heck-it this last little bit, but I plan to pick this up again this weekend or next

**Comment by devlead** [comment]
@JonCubed ok perfect 👍

**Comment by devlead** [comment]
@JonCubed FYI feature branch is now merged into develop I've rebased and retargeted PRs against develop

**Comment by JonCubed** [comment]
@devlead I pulled latest and ran into a lot of package restore issues variants of the following
`Unable to resolve 'Microsoft.TargetingPack.NETFramework.v4.5.2 (>= 1.0.1)' for '.NETFramework,Version=v4.5.2'.`

This package doesn't seem to be available on nuget.org but found it on the [dotnet myget feed](https://dotnet.myget.org/F/dotnet-core/api/v3/index.json) added it to my nuget sources and all is well again. Not sure if this is the correct thing to do to go about fixing it.

**Comment by devlead** [comment]
@JonCubed yes it's not on nuget yet and you found the feed it's on, running build.cake would restore it though 
https://github.com/cake-build/cake/blob/develop/build.cake#L103-L106

The new tooling removed support for compiling to 4.5 on Unix and that's why we currently need it. A nuget.config in repo would assist those not building cake with cake first though.

**Comment by JonCubed** [comment]
@devlead Ah ok. I didn't even think to try using the build script, I was just trying to verify everything compiled in VS.

I also noticed that you have a [global.json](https://github.com/cake-build/cake/blob/develop/global.json#L4) that pins the cli to preview1. Is that intentional? The build.ps1 is using 1.0.4 of the cli so doesn't seems so

**Comment by devlead** [comment]
That should be 1.0.4 yes

**Comment by JonCubed** [comment]
@devlead msbuild branch is ready for a review

**Comment by JonCubed** [comment]
@devlead vstest is ready for review @ #1675. That was the last command to do! Let me know if you need/want anything else done.
